File: leetcode22.cpp
class Solution {
	vector<string> res;
public:
	void generate(int n, int i, int j, string str)
	{
		if(i == n && j == n)
			res.push_back(str);
		if(i < n)
			generate(n,i+1,j,str+'(');
		if(j < n && j < i)
			generate(n,i,j+1,str+')');
	}

    vector<string> generateParenthesis(int n) {
        string str;
        generate(n,0,0,str);
        return res;
    }
};

File: leetcode22.java
public class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> res = new ArrayList<>();
        backtracking(res,"",0,0,n);
        return res;
    }
    
    private void backtracking(List<String> res, String cur, int open, int close, int max){
        // end condition: contains 3 pair of ()
        if(cur.length() == max * 2){
            res.add(cur);
            return;
        }
        

        if(close < open){
            backtracking(res,cur+')',open,close+1,max);
        }
        if(open < max){
            backtracking(res,cur+'(',open+1,close,max);
        }
    }
}
